R/setops.R

```diff
@@ -51,15 +53,22 @@ funique <- function(x) {
   if (any(dup)) .Call(CsubsetDT, x, which_(dup, FALSE), seq_along(x)) else x
 }
 
-fintersect <- function(x, y, all=FALSE) {
+.set_ops_arg_check = function(x, y, all, .seqn = FALSE, block_list = TRUE) {
   if (!is.logical(all) || length(all) != 1L) stop("argument 'all' should be logical of length one")
-  if (!is.data.table(x) || !is.data.table(y)) stop("x and y must be both data.tables")
-  if (!identical(sort(names(x)), sort(names(y)))) stop("x and y must have same column names")
-  if (!identical(names(x), names(y))) stop("x and y must have same column order")
-  bad.type = setNames(c("raw","complex","list") %chin% c(vapply(x, typeof, FUN.VALUE = ""), vapply(y, typeof, FUN.VALUE = "")), c("raw","complex","list"))
-  if (any(bad.type)) stop(sprintf("x and y must not have unsupported column types: %s", paste(names(bad.type)[bad.type], collapse=", ")))
-  if (!identical(lapply(x, class), lapply(y, class))) stop("x and y must have same column classes")
-  if (".seqn" %chin% names(x)) stop("None of the datasets to intersect should contain a column named '.seqn'")
+  if (!is.data.table(x) || !is.data.table(y)) stop("x and y must both be data.tables")
+  if (!identical(sort(names(x)), sort(names(y)))) stop("x and y must have the same column names")
+  if (!identical(names(x), names(y))) stop("x and y must have the same column order")
+  bad_types = c("raw", "complex", if (block_list) "list")
+  found = bad_types %chin% c(vapply(x, typeof, FUN.VALUE = ""),
+                             vapply(y, typeof, FUN.VALUE = ""))
+  if (any(found)) stop("unsupported column type", if (sum(found) > 1L) "s" else "",
+                       " found in x or y: ", brackify(bad_types[found]))
+  if (!identical(lapply(x, class), lapply(y, class))) stop("x and y must have the same column classes")
+  if (.seqn && ".seqn" %chin% names(x)) stop("None of the datasets should contain a column named '.seqn'")
+}
+
+fintersect <- function(x, y, all=FALSE) {
+  .set_ops_arg_check(x, y, all, .seqn = TRUE)
```

Comment:
This won't impact performance right? Won't be copying `x`/`y`?

Reply:
no